File: test/v.3.33.33/ProjectTemplate/02. Console Project/vs2019/Dev.DevKit.Shared/Lib/Date.cs
using System.Diagnostics;
using System.Globalization;
using System.Runtime.Serialization;

namespace System
{
    [DebuggerNonUserCode()]
    public struct Date : IComparable, IFormattable, IComparable<Date>, IEquatable<Date>
    {
        private DateTime _dt;

        public static readonly Date MaxValue = new Date(DateTime.MaxValue);
        public static readonly Date MinValue = new Date(DateTime.MinValue);

        public Date(int year, int month, int day)
        {
            this._dt = new DateTime(year, month, day);
        }

        public Date(DateTime dateTime)
        {
            this._dt = dateTime.AddTicks(-dateTime.Ticks % TimeSpan.TicksPerDay);
        }

        public static TimeSpan operator -(Date d1, Date d2)
        {
            return d1._dt - d2._dt;
        }

        public static Date operator -(Date d, TimeSpan t)
        {
            return new Date(d._dt - t);
        }

        public static bool operator !=(Date d1, Date d2)
        {
            return d1._dt != d2._dt;
        }

        public static Date operator +(Date d, TimeSpan t)
        {
            return new Date(d._dt + t);
        }

        public static bool operator <(Date d1, Date d2)
        {
            return d1._dt < d2._dt;
        }

        public static bool operator <=(Date d1, Date d2)
        {
            return d1._dt <= d2._dt;
        }

        public static bool operator ==(Date d1, Date d2)
        {
            return d1._dt == d2._dt;
        }

        public static bool operator >(Date d1, Date d2)
        {
            return d1._dt > d2._dt;
        }

        public static bool operator >=(Date d1, Date d2)
        {
            return d1._dt >= d2._dt;
        }

        public static implicit operator DateTime(Date d)
        {
            return d._dt;
        }

        public static explicit operator Date(DateTime d)
        {
            return new Date(d);
        }

        public int Day
        {
            get
            {
                return this._dt.Day;
            }
        }

        public DayOfWeek DayOfWeek
        {
            get
            {
                return this._dt.DayOfWeek;
            }
        }

        public int DayOfYear
        {
            get
            {
                return this._dt.DayOfYear;
            }
        }

        public int Month
        {
            get
            {
                return this._dt.Month;
            }
        }

        public static Date Today
        {
            get
            {
                return new Date(DateTime.Today);
            }
        }

        public int Year
        {
            get
            {
                return this._dt.Year;
            }
        }

        public Date AddDays(int value)
        {
            return new Date(this._dt.AddDays(value));
        }

        public Date AddMonths(int value)
        {
            return new Date(this._dt.AddMonths(value));
        }

        public Date AddYears(int value)
        {
            return new Date(this._dt.AddYears(value));
        }

        public static int Compare(Date d1, Date d2)
        {
            return d1.CompareTo(d2);
        }

        public int CompareTo(Date value)
        {
            return this._dt.CompareTo(value._dt);
        }

        public int CompareTo(object value)
        {
            return this._dt.CompareTo(value);
        }

        public static int DaysInMonth(int year, int month)
        {
            return DateTime.DaysInMonth(year, month);
        }

        public bool Equals(Date value)
        {
            return this._dt.Equals(value._dt);
        }

        public override bool Equals(object value)
        {
            return value is Date && this._dt.Equals(((Date)value)._dt);
        }

        public override int GetHashCode()
        {
            return this._dt.GetHashCode();
        }

        public static bool Equals(Date d1, Date d2)
        {
            return d1._dt.Equals(d2._dt);
        }

        public static bool IsLeapYear(int year)
        {
            return DateTime.IsLeapYear(year);
        }

        public static Date Parse(string s)
        {
            return new Date(DateTime.Parse(s));
        }

        public static Date Parse(string s, IFormatProvider provider)
        {
            return new Date(DateTime.Parse(s, provider));
        }

        public static Date Parse(string s, IFormatProvider provider, DateTimeStyles style)
        {
            return new Date(DateTime.Parse(s, provider, style));
        }

        public static Date ParseExact(string s, string format, IFormatProvider provider)
        {
            return new Date(DateTime.ParseExact(s, format, provider));
        }

        public static Date ParseExact(string s, string format, IFormatProvider provider, DateTimeStyles style)
        {
            return new Date(DateTime.ParseExact(s, format, provider, style));
        }

        public static Date ParseExact(string s, string[] formats, IFormatProvider provider, DateTimeStyles style)
        {
            return new Date(DateTime.ParseExact(s, formats, provider, style));
        }

        public TimeSpan Subtract(Date value)
        {
            return this - value;
        }

        public Date Subtract(TimeSpan value)
        {
            return this - value;
        }

        public string ToLongString()
        {
            return this._dt.ToLongDateString();
        }

        public string ToShortString()
        {
            return this._dt.ToShortDateString();
        }

        public override string ToString()
        {
            return this.ToShortString();
        }

        public string ToString(IFormatProvider provider)
        {
            return this._dt.ToString(provider);
        }

        public string ToString(string format)
        {
            if (format == "O" || format == "o" || format == "s")
            {
                return this.ToString("yyyy-MM-dd");
            }

            return this._dt.ToString(format);
        }

        public string ToString(string format, IFormatProvider provider)
        {
            return this._dt.ToString(format, provider);
        }

        public static bool TryParse(string s, out Date result)
        {
            DateTime d;
            bool success = DateTime.TryParse(s, out d);
            result = new Date(d);
            return success;
        }

        public static bool TryParse(string s, IFormatProvider provider, DateTimeStyles style, out Date result)
        {
            DateTime d;
            bool success = DateTime.TryParse(s, provider, style, out d);
            result = new Date(d);
            return success;
        }

        public static bool TryParseExact(string s, string format, IFormatProvider provider, DateTimeStyles style, out Date result)
        {
            DateTime d;
            bool success = DateTime.TryParseExact(s, format, provider, style, out d);
            result = new Date(d);
            return success;
        }

        public static bool TryParseExact(string s, string[] formats, IFormatProvider provider, DateTimeStyles style, out Date result)
        {
            DateTime d;
            bool success = DateTime.TryParseExact(s, formats, provider, style, out d);
            result = new Date(d);
            return success;
        }
    }

    [DebuggerNonUserCode()]
    public static class DateTimeExtensions
    {
        public static Date ToDate(this DateTime dt)
        {
            return new Date(dt);
        }

        public static DateTime ToDateTime(this Date d)
        {
            return new DateTime(d.Year, d.Month, d.Day, 0, 0, 0);
        }
    }
}

File: v5/ProjectTemplates/VisualBasic/01.SharedProjectTemplate/Lib/Date.vb
Imports System
Imports System.Diagnostics
Imports System.Globalization

Namespace System
    <DebuggerNonUserCode()>
    Public Structure [Date]
        Implements IComparable
        Implements IFormattable
        Implements IComparable(Of [Date])
        Implements IEquatable(Of [Date])

        Private _dt As DateTime

        Public Shared ReadOnly MaxValue As New [Date](DateTime.MaxValue)
        Public Shared ReadOnly MinValue As New [Date](DateTime.MinValue)

        Public Sub New(ByVal year As Integer, ByVal month As Integer, ByVal day As Integer)
            Me._dt = New DateTime(year, month, day)
        End Sub

        Public Sub New(ByVal dateTime As DateTime)
            Me._dt = dateTime.AddTicks(-dateTime.Ticks Mod TimeSpan.TicksPerDay)
        End Sub

        Public Shared Operator -(ByVal d1 As [Date], ByVal d2 As [Date]) As TimeSpan
            Return d1._dt - d2._dt
        End Operator

        Public Shared Operator -(ByVal d As [Date], ByVal t As TimeSpan) As [Date]
            Return New [Date](d._dt - t)
        End Operator

        Public Shared Operator <>(ByVal d1 As [Date], ByVal d2 As [Date]) As Boolean
            Return d1._dt <> d2._dt
        End Operator

        Public Shared Operator +(ByVal d As [Date], ByVal t As TimeSpan) As [Date]
            Return New [Date](d._dt + t)
        End Operator

        Public Shared Operator <(ByVal d1 As [Date], ByVal d2 As [Date]) As Boolean
            Return d1._dt < d2._dt
        End Operator

        Public Shared Operator <=(ByVal d1 As [Date], ByVal d2 As [Date]) As Boolean
            Return d1._dt <= d2._dt
        End Operator

        Public Shared Operator =(ByVal d1 As [Date], ByVal d2 As [Date]) As Boolean
            Return d1._dt = d2._dt
        End Operator

        Public Shared Operator >(ByVal d1 As [Date], ByVal d2 As [Date]) As Boolean
            Return d1._dt > d2._dt
        End Operator

        Public Shared Operator >=(ByVal d1 As [Date], ByVal d2 As [Date]) As Boolean
            Return d1._dt >= d2._dt
        End Operator

        Public Shared Widening Operator CType(ByVal d As [Date]) As DateTime
            Return d._dt
        End Operator

        Public Shared Narrowing Operator CType(ByVal d As DateTime) As [Date]
            Return New [Date](d)
        End Operator

        Public ReadOnly Property Day As Integer
            Get
                Return Me._dt.Day
            End Get
        End Property

        Public ReadOnly Property DayOfWeek As DayOfWeek
            Get
                Return Me._dt.DayOfWeek
            End Get
        End Property

        Public ReadOnly Property DayOfYear As Integer
            Get
                Return Me._dt.DayOfYear
            End Get
        End Property

        Public ReadOnly Property Month As Integer
            Get
                Return Me._dt.Month
            End Get
        End Property

        Public Shared ReadOnly Property Today As [Date]
            Get
                Return New [Date](DateTime.Today)
            End Get
        End Property

        Public ReadOnly Property Year As Integer
            Get
                Return Me._dt.Year
            End Get
        End Property

        Public Function AddDays(ByVal value As Integer) As [Date]
            Return New [Date](Me._dt.AddDays(value))
        End Function

        Public Function AddMonths(ByVal value As Integer) As [Date]
            Return New [Date](Me._dt.AddMonths(value))
        End Function

        Public Function AddYears(ByVal value As Integer) As [Date]
            Return New [Date](Me._dt.AddYears(value))
        End Function

        Public Shared Function Compare(ByVal d1 As [Date], ByVal d2 As [Date]) As Integer
            Return d1.CompareTo(d2)
        End Function

        Public Function CompareTo(ByVal value As [Date]) As Integer Implements IComparable(Of [Date]).CompareTo
            Return Me._dt.CompareTo(value._dt)
        End Function

        Public Function CompareTo(ByVal value As Object) As Integer Implements IComparable.CompareTo
            Return Me._dt.CompareTo(value)
        End Function

        Public Shared Function DaysInMonth(ByVal year As Integer, ByVal month As Integer) As Integer
            Return DateTime.DaysInMonth(year, month)
        End Function

        Public Overloads Function Equals(ByVal value As [Date]) As Boolean Implements IEquatable(Of [Date]).Equals
            Return Me._dt.Equals(value._dt)
        End Function

        Public Overrides Function Equals(ByVal value As Object) As Boolean
            If TypeOf value Is [Date] Then
                Return Me._dt.Equals(CType(value, [Date])._dt)
            End If
            Return False
        End Function

        Public Overrides Function GetHashCode() As Integer
            Return Me._dt.GetHashCode()
        End Function

        Public Shared Overloads Function Equals(ByVal d1 As [Date], ByVal d2 As [Date]) As Boolean
            Return d1._dt.Equals(d2._dt)
        End Function

        Public Shared Function IsLeapYear(ByVal year As Integer) As Boolean
            Return DateTime.IsLeapYear(year)
        End Function

        Public Shared Function Parse(ByVal s As String) As [Date]
            Return New [Date](DateTime.Parse(s))
        End Function

        Public Shared Function Parse(ByVal s As String, ByVal provider As IFormatProvider) As [Date]
            Return New [Date](DateTime.Parse(s, provider))
        End Function

        Public Shared Function Parse(ByVal s As String, ByVal provider As IFormatProvider, ByVal style As DateTimeStyles) As [Date]
            Return New [Date](DateTime.Parse(s, provider, style))
        End Function

        Public Shared Function ParseExact(ByVal s As String, ByVal format As String, ByVal provider As IFormatProvider) As [Date]
            Return New [Date](DateTime.ParseExact(s, format, provider))
        End Function

        Public Shared Function ParseExact(ByVal s As String, ByVal format As String, ByVal provider As IFormatProvider, ByVal style As DateTimeStyles) As [Date]
            Return New [Date](DateTime.ParseExact(s, format, provider, style))
        End Function

        Public Shared Function ParseExact(ByVal s As String, ByVal formats As String(), ByVal provider As IFormatProvider, ByVal style As DateTimeStyles) As [Date]
            Return New [Date](DateTime.ParseExact(s, formats, provider, style))
        End Function

        Public Function Subtract(ByVal value As [Date]) As TimeSpan
            Return Me - value
        End Function

        Public Function Subtract(ByVal value As TimeSpan) As [Date]
            Return Me - value
        End Function

        Public Function ToLongString() As String
            Return Me._dt.ToLongDateString()
        End Function

        Public Function ToShortString() As String
            Return Me._dt.ToShortDateString()
        End Function

        Public Overrides Function ToString() As String
            Return Me.ToShortString()
        End Function

        Public Overloads Function ToString(ByVal provider As IFormatProvider) As String Implements IFormattable.ToString
            Return Me._dt.ToString(provider)
        End Function

        Public Overloads Function ToString(ByVal format As String) As String
            If format = "O" OrElse format = "o" OrElse format = "s" Then
                Return Me.ToString("yyyy-MM-dd")
            End If

            Return Me._dt.ToString(format)
        End Function

        Public Overloads Function ToString(ByVal format As String, ByVal provider As IFormatProvider) As String Implements IFormattable.ToString
            Return Me._dt.ToString(format, provider)
        End Function

        Public Shared Function TryParse(ByVal s As String, ByRef result As [Date]) As Boolean
            Dim d As DateTime
            Dim success As Boolean = DateTime.TryParse(s, d)
            result = New [Date](d)
            Return success
        End Function

        Public Shared Function TryParse(ByVal s As String, ByVal provider As IFormatProvider, ByVal style As DateTimeStyles, ByRef result As [Date]) As Boolean
            Dim d As DateTime
            Dim success As Boolean = DateTime.TryParse(s, provider, style, d)
            result = New [Date](d)
            Return success
        End Function

        Public Shared Function TryParseExact(ByVal s As String, ByVal format As String, ByVal provider As IFormatProvider, ByVal style As DateTimeStyles, ByRef result As [Date]) As Boolean
            Dim d As DateTime
            Dim success As Boolean = DateTime.TryParseExact(s, format, provider, style, d)
            result = New [Date](d)
            Return success
        End Function

        Public Shared Function TryParseExact(ByVal s As String, ByVal formats As String(), ByVal provider As IFormatProvider, ByVal style As DateTimeStyles, ByRef result As [Date]) As Boolean
            Dim d As DateTime
            Dim success As Boolean = DateTime.TryParseExact(s, formats, provider, style, d)
            result = New [Date](d)
            Return success
        End Function
    End Structure

    <DebuggerNonUserCode()>
    Public Module DateTimeExtensions
        <Runtime.CompilerServices.Extension()>
        Public Function ToDate(ByVal dt As DateTime) As [Date]
            Return New [Date](dt)
        End Function

        <Runtime.CompilerServices.Extension()>
        Public Function ToDateTime(ByVal d As [Date]) As DateTime
            Return New DateTime(d.Year, d.Month, d.Day, 0, 0, 0)
        End Function
    End Module
End Namespace
